### ISSUE TITLE
Chunk `TAX_IDTAXA` input to speed up taxonomic assignment

### ISSUE THREAD

**Issue by jackscanlan** [AVR-biosecurity-bioinformatics/freyr#16]
Big bottleneck at the moment with larger datasets is `TAX_IDTAXA`, which can only run with a single core and time required scales with the combination of # of input sequences and size of the reference database. 

One way around this could be to split the input sequence table into chunks of fixed size (perhaps 100 sequences?), then run those through `TAX_IDTAXA`, combining them at the end. This would require new `TAX_CHUNK` (?) and `TAX_COMBINE` (?) modules to bookend the existing module.

**Comment by alexpiper** [comment]
Bumping this issue as it seems to be a particular problem for diverse eDNA and microbiome datasets. 

On a single miseq run the `TAX_IDTAXA` step accounts for ~70% of the pipeline runtime and the process can run over the 6hr limit of the `shortrun` partition, causing it to fail.

**Comment by jackscanlan** [comment]
I'll make this a priority to look into on Monday, @alexpiper

**Comment by alexpiper** [comment]
Okay after running this dataset again it looks like the failures are more of a process time limit issue, rather than shortrun limits. Although shortrun limits could become an issue if analysing more than one miseq run of this data type.

The dataset has 12,520 filtered ASV's going into IDTAXA, see working directory `/group/pathogens/IAWS/Personal/Alexp/metabarcoding/test`

TAX_IDTAXA currently has the 'medium' resources which has a 2hr time limit:
```
withLabel: medium {
        cpus    = { check_max( 1 , 'cpus'    ) }
        memory  = { check_max( 8.GB * task.attempt, 'memory'  ) }
        time    = { check_max( 2.h * task.attempt, 'time'     ) }
    }
```
Attempt 1 timed out at 2hr
Attempt 2 timed out at 4hr
Attempt 3 completes sucessfully at 4h 7m

An interim solution could be to increase the time limit for this specific process.

**Comment by alexpiper** [comment]
On a larger dataset of the same samples (23,000 ASVS) TAX_IDTAXA runs for 6hrs on the `shortrun` partition then times out before correctly switching to `batch` partition and running for 8hrs 7 mins before completing successfully. The whole run completes in 14hrs 51 minutes with IDTAXA accounting for 14hrs 7 mins (95% of runtime).

This is using the terrestrial arthropod reference database and seems to be due to the number of ASVS and the size of the databases.  The marine metabarcoding eDNA datasets take ~15mins to run per flowcell with a similar number of ASVS due to the much smaller reference database.

**Comment by jackscanlan** [comment]
This is going to require some reworking of the way the pipeline passes sequences between processes in the `TAXONOMY` subworkflow -- at the moment it uses matrices of sample ID x sequence string, but these can't be split with Nextflow operators: we need a tabular format with one row per sequence. 

`TAX_IDTAXA` also has some counterintuitive outputs -- `tax`, which is an `.rds` tibble of taxonomic assignments, and `ids`, which is an `.rds` of IDTAXA confidences + taxonomy of class `Taxa`. Neither of these outputs have sequence ID information in them, relying on order to join back to sequences in subsequent processes. 

Ideally sequences should be passed between processes as `.fasta` (there is a small amount of overhead converting these between formats in R, but it makes the pipeline easier to debug), and taxonomic information should be passed as tabular files (`.csv` probably), except where absolutely necessary.

**Comment by alexpiper** [comment]
I agree with these changes - the taxonomy modules dont use count information so a fasta input makes sense. And if the ASV fasta files from multiple flowcells are merged and deduplicated before going into the respective taxonomy module it would avoid assigning the same sequence more than once and then having to check for duplicates afterwards as we do in `MERGE_TAX`

And standardising outputs from the taxonomy modules will definitely help when thinking about consensus/ensemble assignments ala https://github.com/AVR-biosecurity-bioinformatics/freyr/issues/64
